File: WebSite/site/scripts/check_scorched_tanks_pwa.py
from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1] / "static" / "play" / "scorched-tanks"


def read_text(path: str) -> str:
    return (ROOT / path).read_text(encoding="utf-8")


def assert_condition(condition: bool, message: str) -> None:
    if not condition:
        raise AssertionError(message)


def main() -> None:
    html = read_text("index.html")
    original = read_text("original.js")
    service_worker = read_text("service-worker.js")
    manifest = json.loads(read_text("manifest.webmanifest"))

    assert_condition(
        'id="install-button"' in html
        and 'id="install-button" class="button" type="button" disabled' not in html,
        "Install button must not be disabled while waiting for beforeinstallprompt.",
    )
    assert_condition(
        "Use the browser menu to install this app" in original,
        "Install fallback must tell users how to finish PWA installation.",
    )
    assert_condition(
        "installButton.disabled = false;" in original,
        "Install button must be re-enabled after native install prompt handling.",
    )
    assert_condition(
        manifest["display"] == "standalone"
        and manifest["start_url"] == "/play/scorched-tanks/"
        and manifest["scope"] == "/play/scorched-tanks/",
        "Manifest must remain installable as a scoped standalone PWA.",
    )
    assert_condition(
        '"./manifest.webmanifest?v=bug043"' in service_worker
        and '"./original.js?v=bug043"' in service_worker,
        "Service worker cache must include the current PWA install assets.",
    )


if __name__ == "__main__":
    main()

File: WebSite/site/scripts/check-scorched-tanks-pwa.mjs
import { readFileSync } from "node:fs";
import { dirname } from "node:path";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const page = readFileSync(
  resolve(root, "static", "play", "scorched-tanks", "index.html"),
  "utf8",
);
const script = readFileSync(
  resolve(root, "static", "play", "scorched-tanks", "original.js"),
  "utf8",
);
const manifest = JSON.parse(
  readFileSync(
    resolve(root, "static", "play", "scorched-tanks", "manifest.webmanifest"),
    "utf8",
  ),
);

function assert(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

assert(
  page.includes('<link rel="manifest" href="./manifest.webmanifest'),
  "Scorched Tanks page must expose a web app manifest.",
);
assert(
  !/<button[^>]+id="install-button"[^>]+disabled/.test(page),
  "Install button must stay clickable when beforeinstallprompt is unavailable.",
);
assert(
  script.includes("beforeinstallprompt"),
  "Install flow must still use the native browser prompt when available.",
);
assert(
  script.includes("Install from this browser"),
  "Install flow must report the manual browser-install path as a fallback.",
);
assert(
  script.includes("serviceWorker") &&
    script.includes('register("./service-worker.js", { scope: "./" })'),
  "Scorched Tanks page must register its local service worker.",
);
assert(
  manifest.display === "standalone" &&
    manifest.start_url === "/play/scorched-tanks/" &&
    manifest.scope === "/play/scorched-tanks/",
  "Manifest must remain browser-installable for the Scorched Tanks route.",
);

console.log("Scorched Tanks PWA contract ok");
